presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java

```diff
@@ -0,0 +1,560 @@
+        {
+            super(semiJoin());
+        }
+
+        @Override
+        protected Result pushDownDereferences(Context context, SemiJoinNode semiJoinNode, BiMap<DereferenceExpression, VariableReferenceExpression> expressions)
+        {
+            Assignments.Builder filteringSourceDereferences = Assignments.builder();
+            Assignments.Builder sourceDereferences = Assignments.builder();
+
+            for (Map.Entry<VariableReferenceExpression, DereferenceExpression> entry : expressions.inverse().entrySet()) {
+                VariableReferenceExpression baseSymbol = getBase(entry.getValue(), context.getVariableAllocator().getTypes());
+                if (semiJoinNode.getFilteringSource().getOutputVariables().contains(baseSymbol)) {
+                    filteringSourceDereferences.put(entry.getKey(), castToRowExpression(entry.getValue()));
+                }
+                else {
+                    sourceDereferences.put(entry.getKey(), castToRowExpression(entry.getValue()));
+                }
+            }
+            PlanNode filteringSource = createProjectBelow(semiJoinNode.getFilteringSource(), filteringSourceDereferences.build(), context.getIdAllocator());
+            PlanNode source = createProjectBelow(semiJoinNode.getSource(), sourceDereferences.build(), context.getIdAllocator());
+            return Result.ofPlanNode(semiJoinNode.replaceChildren(ImmutableList.of(source, filteringSource)));
+        }
+    }
+
+    public class PushDownDereferenceThroughProject
+            extends PushdownDereferencesInProject<ProjectNode>
+    {
+        PushDownDereferenceThroughProject()
+        {
+            super(project());
+        }
+
+        @Override
+        protected Result pushDownDereferences(Context context, ProjectNode projectNode, BiMap<DereferenceExpression, VariableReferenceExpression> expressions)
+        {
+            ImmutableMap<VariableReferenceExpression, RowExpression>
+                    dereferencesMap =
+                    expressions.inverse().entrySet().stream().collect(toImmutableMap(Map.Entry::getKey,
+                            entry -> castToRowExpression(entry.getValue())));
+
+            return Result.ofPlanNode(
+                    new ProjectNode(context.getIdAllocator().getNextId(),
+                            projectNode.getSource(),
+                            Assignments.builder().putAll(projectNode.getAssignments()).putAll(dereferencesMap).build()));
+        }
+    }
+
+    public class PushDownDereferenceThroughUnnest
+            extends PushdownDereferencesInProject<UnnestNode>
+    {
+        PushDownDereferenceThroughUnnest()
+        {
+            super(unnest());
+        }
+
+        @Override
+        protected Result pushDownDereferences(Context context, UnnestNode unnestNode, BiMap<DereferenceExpression, VariableReferenceExpression> expressions)
+        {
+            // Create new Project contains all pushdown symbols above original source
```

Comment:
- remove comments, these don't seem to provide additional information 
- format for readability and consistency

```
            Map<VariableReferenceExpression, RowExpression> dereferencesMap = expressions.inverse().entrySet().stream()
                    .collect(toImmutableMap(Map.Entry::getKey, entry -> castToRowExpression(entry.getValue())));
            Assignments assignments = Assignments.builder()
                    .putAll(identityAssignmentsAsSymbolReferences(unnestNode.getSource().getOutputVariables()))
                    .putAll(dereferencesMap)
                    .build();
            ProjectNode source = new ProjectNode(context.getIdAllocator().getNextId(), unnestNode.getSource(), assignments);

            return Result.ofPlanNode(
                    new UnnestNode(context.getIdAllocator().getNextId(),
                            source,
                            ImmutableList.<VariableReferenceExpression>builder()
                                    .addAll(unnestNode.getReplicateVariables())
                                    .addAll(expressions.values())
                                    .build(),
                            unnestNode.getUnnestVariables(),
                            unnestNode.getOrdinalityVariable()));
```

Reply:
fixed